### PR TITLE
rgw: migrate atomic_t to std::atomic<> (ebirah)

### PR DESCRIPTION
**Pull request by chardan**
Refer-to: https://github.com/ceph/ceph/pull/14655

This should contain all RGW changes (it's possible that there are minor patches elsewhere, but this will certainly be the bulk of them).

Signed-off-by: Jesse Williamson <jwilliamson@suse.de>

### CODE REVIEW

**Comment by chardan** [comment]
@cbodley

File: src/rgw/rgw_coroutine.h [hunk 1]
```diff
@@ -542,7 +544,7 @@ class RGWCoroutinesManager {
   int run(list<RGWCoroutinesStack *>& ops);
   int run(RGWCoroutine *op);
   void stop() {
-    if (going_down.inc() == 1) {
+    if (!going_down) {
```
Comment (cbodley) [inline src/rgw/rgw_coroutine.h]: the `inc()` here was important, allowing `stop()` to be called multiple times, but only calling `completion_mgr->go_down()` the first time

you -could- turn it back into an `atomic<unsigned>` and use `if (++going_down == 1)`, but i think [compare_exchange](http://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange) is a better fit for this:
```c++
bool expected = false;
if (going_down.compare_exchange_strong(expected, true)) {
  completion_mgr->go_down();
}
```

Comment (chardan) [inline src/rgw/rgw_coroutine.h]: Good catch! Thank you. Patched.

**Comment by cbodley** [comment]
after fix to `RGWCoroutinesManager::stop()`, passed rgw suite: http://pulpito.ceph.com/cbodley-2017-04-27_23:02:56-rgw-wip-cbodley-testing---basic-mira/

**Pull request merged**